revivelib/core.go

```diff
@@ -0,0 +1,210 @@
+package revivelib
+
+import (
+	"io/ioutil"
+	"log"
+	"strings"
+
+	"github.com/mgechev/dots"
+	"github.com/mgechev/revive/config"
+	"github.com/mgechev/revive/lint"
+	"github.com/mgechev/revive/logging"
+	"github.com/pkg/errors"
+)
+
+// Revive is responsible for running linters and formatters
+// and returning a set of results.
+type Revive struct {
+	config          *lint.Config
+	lintingRules    []lint.Rule
+	logger          *log.Logger
+	maxOpenFiles    int
+	excludePatterns ArrayFlags
+}
+
+// New creates a new instance of Revive lint runner.
+func New(
+	conf *lint.Config,
+	setExitStatus bool,
+	maxOpenFiles int,
+	excludePatterns ArrayFlags,
+	extraRules ...ExtraRule,
+) (*Revive, error) {
+	log, err := logging.GetLogger()
+	if err != nil {
+		return nil, errors.Wrap(err, "initializing revive - getting logger")
+	}
+
+	if setExitStatus {
+		conf.ErrorCode = 1
+		conf.WarningCode = 1
+	}
+
+	extraRuleInstances := make([]lint.Rule, len(extraRules))
+	for i, extraRule := range extraRules {
+		extraRuleInstances[i] = extraRule.Rule
+
+		ruleName := extraRule.Rule.Name()
+
+		_, isRuleAlreadyConfigured := conf.Rules[ruleName]
+		if !isRuleAlreadyConfigured {
+			conf.Rules[ruleName] = extraRule.DefaultConfig
+		}
+	}
+
+	lintingRules, err := config.GetLintingRules(conf, extraRuleInstances)
+	if err != nil {
+		return nil, errors.Wrap(err, "initializing revive - gettint lint rules")
+	}
+
+	log.Println("Config loaded")
+
+	if len(excludePatterns) == 0 { // if no excludes were set in the command line
+		excludePatterns = conf.Exclude // use those from the configuration
+	}
+
+	return &Revive{
+		logger:          log,
+		config:          conf,
+		lintingRules:    lintingRules,
+		maxOpenFiles:    maxOpenFiles,
+		excludePatterns: excludePatterns,
+	}, nil
+}
+
+// Lint files in the specified paths.
+func (r *Revive) Lint(includePatterns ...string) (<-chan lint.Failure, error) {
+	packages, err := getPackages(includePatterns, r.excludePatterns)
+	if err != nil {
+		return nil, errors.Wrap(err, "linting - getting packages")
+	}
+
+	revive := lint.New(func(file string) ([]byte, error) {
+		contents, err := ioutil.ReadFile(file)
+
+		if err != nil {
+			return nil, errors.Wrap(err, "reading file "+file)
+		}
+
+		return contents, nil
+	}, r.maxOpenFiles)
+
+	failures, err := revive.Lint(packages, r.lintingRules, *r.config)
+	if err != nil {
+		return nil, errors.Wrap(err, "linting - retrieving failures channel")
+	}
+
+	return failures, nil
+}
+
+// GetLintFailures gets the list of failures for a given failures channel from Lint.
+func (r *Revive) GetLintFailures(failuresChan <-chan lint.Failure) []lint.Failure {
```

Comment:
Not sure about the added value of this method.
We do not use it and other users of the lib can implement it very easily.

Reply:
I wasn't sure either, but feels like it's a good helper method. If you prefer I can remove.